### DISCUSSION TITLE
Server-Side Code Isolation in Start

### DISCUSSION THREAD

**Discussion by thefrana** [TanStack/router#2671] [Q&A] [answered]
First,
I wanted to express my appreciation for these excellent packages.

I have a question regarding server-side code isolation: Is there a mechanism to ensure certain code runs exclusively on the server? For context:
- Remix uses a `.server` file naming convention to automatically exclude server code from the client bundle
- While createServerFn works well for simple functions, it's not ideal for more complex server modules (e.g., those requiring database connections)

Is there a recommended approach or best practice for handling this use case?

**Comment by ericwilhite** [comment, marked as answer]
I am experimenting with Start and had the same concerns about leaking sensitive things like db/api credentials and connections. 

Couple of options I've found that are worth a test:

- [`import 'server-only'`](https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#keeping-server-only-code-out-of-the-client-environment): I don't particularly like this method because I don't want credentials leaking because of me forgetting to put the import in. 

- [`vite-env-only`](https://github.com/pcattori/vite-env-only): I've not tested this but by the looks of it I'm really hoping it works. I think this + an `eslint` rule to ensure `createsServerFn` calls only exist in a `**/*.server.*` pattern is going to be my preferred path forward.

**Reply by thefrana** [reply]
Thanks for the answer, we use `vite-env-only` extensively in other projects, but the TypeScript makes it return the actual value or undefined, which breaks normal flow. Also, the syntax is awful when you need to wrap **everything** in `serverOnly$` macro.

We will try `import 'server-only'`. Thanks for that.

**Reply by ericwilhite** [reply]
It was my understanding with `vite-env-only` you could define everything within the vite plugin and not worry about using the server/client macros they expose?

I plan on testing it out next week using the `*.server.ts` pattern I mentioned.

**Reply by thefrana** [reply]
Yeah, you are absolutely right. I missed the option in the docs. Thanks for noticing!
